[snmp] adding check

truthbk opened the pull request:
### What does this PR do?

Adds the SNMP check to agent6. Note that it depends on the net-snmp libraries being available on the host.

### Motivation

We're going to need this check in agent 6. It was about time!


### Testing Guidelines

Added pertinent tests. We do not actually test all functionality, the snmpGet function is a little hard to test given its IO nature. We have "skipped" it in the tests and provide fake varbind results in its place. Also pending are "table" varbind results, these are a little harder to mock so I've left them for a future implementation.

Added `snmp` dependencies to CircleCI.

### Additional Notes

We should discuss InitConfig, I am currently ignoring it and just going with the defaults. The reason being that the snmp library can only really be initialized once, so we have to gauge if we should be enforcing this kind of thing in the core, or in the check, there should be no issues with the second approach.

An elegant failure mechanism should be put in place if the snmp libraries are not available.

truthbk committed:
[snmp] adding snmp check (WIP).

[misc] compile fixes.

[build][osx] build on OSX without the agent installed.

[snmp][glide] adding deps.

[python][osx] use tweaked header file.

[snmp] updating interface.

[snmp] fix build issues.

[snmp] generate translator and tagmap when the check is configured O(1).

[snmp] changes to build on linux and get config to load.

[snmp] parse config unmarshal with instance pointer.

[snmp] use factory when registering check.

[snmp] fixing multiple lint issues.

[snmp][test][wip] adding snmp tests.

[snmp][test][wip] wrong name for test - fixing.

[util][bimap] fixing lint issues + tests.

[tests][core] moving common testing facilities to its own package.

[check][testing] fixing lint issues for common. Changing filename.

[core][tests] fixing tests to use common check facilities.

[tests][common] removing file, moved elsewhere.

[snmp] miscellaneous fixes, also to tests.

[snmp][tests] tests must be exported to actually run :facepalm:

[snmp] port is an optional argument, default to 161.

[snmp][test] initialize properly before calling the relevant functions.

[snmp] commit submission, support more metric types.

[snmp][test] fix the tests. Table varbinds are hard - keeping test simple for now.

[snmp] fixing multiple style issues

[travis] add snmp dependencies to the build.

[snmp] graceful failure w/out SNMP libs + initConfig.

[snmp] use sync.Once to initialize SNMP. Fix a couple linting issues.

[pkg-config] removing odd line.

[os x] if we don't have agent, use bundle python headers.

[osx] use pkg-config specified includes. Remove abomination.

masci reviewed (changes requested):
I didn't go deeply the SNMP implementation b/c I don't have the expertise, feel free to ping someone for further review!

olivielpeau reviewed (commented):
Only had a very quick look at the logic but overall this looks good. I made a few comments here and there (many of them are open questions)

truthbk committed:
[snmp] use unique identifier for check instances.

truthbk committed:
[snmp] remove hard mib path, its configurable

truthbk committed:
[snmp][skip ci] making cgo directives more readable.

truthbk committed:
[test] adding reqs to gitlab, setting right testing packages.

[gitlab] add non-free repos..

[snmp][test] fixing test vetting.

[glide] updating lockfile.

[gitlab] missing devel package for net-snmp.

[test][lint] fixing lint issue.

[snmp] provide build dependency.

[snmp] add consts.

[snmp] fix typo.

truthbk committed:
[tests][common] adding comments to exported funcs.

olivielpeau reviewed (approved):
LGTM!

Feel free to merge once these are 👍 :

* There are some merge conflicts :(
* The omnibus build is failing on a health check for the netsnmp lib, could you have a look? https://gitlab.datad0g.com/datadog/datadog-agent/pipelines/44368
* Let's not forget to switch the default `omnibus-software` branch back to `master`

truthbk committed:
[snmp][tests][lint] adding comment for service check.

[snmp][test] restore mocked commit call.

truthbk committed:
[snmp] add embedded location to cflags.

truthbk committed:
[snmp] switch to net-snmp-slim dependency (essentials).

truthbk committed:
[snmp][cgo] build against embedded deps/includes first, left to right precedence.

truthbk committed:
[omnibus] pass LD_RUN_PATH to rake build + rebase fixes.

truthbk committed:
[snmp] definition has been renamed to net-snmp-lib, now on master.

truthbk committed:
[collector] consolidate includes with master.
